### PR TITLE
Fix JS glue-code so that its variables doesn't leak.

### LINKED ISSUE #84: JavaScript - cannot use "fs" without importing it under another name
Essentially, in problems where you need to read from a file, the typical thing to do in node.js is something like `const fs = require("fs");`. Unfortunately, doing this will give an error - the glue code use to run the user code defines `fs`, so node will cause an error when this line is included in user code.

```
Error: Traceback (most recent call last):
  File "/engine/engine/api.py", line 182, in on_post
    self.container_name, code_filename, function_name, input_tuples, output_tuples
  File "/engine/engine/code_runner.py", line 137, in run
    raise EngineExecutionError(exec_stdout)
engine.code_runner.EngineExecutionError: /root/391965dbcab4a713e866ccfc5768821b270dcda8.js:63
    var fs = require('fs');
        ^

SyntaxError: Identifier 'fs' has already been declared
    at Module._compile (internal/modules/cjs/loader.js:723:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
Traceback (most recent call last):
  File "/root/391965dbcab4a713e866ccfc5768821b270dcda8.run.py", line 48, in 
    with open(output_json, mode='r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '391965dbcab4a713e866ccfc5768821b270dcda8.output0.json'
```

Using `fs` without declaring it (just using `fs.whatever()`) won't work either, since it's undefined at the time when the user code is called (despite the variable being hoisted, allowing it to be used as an identifier). This means the user needs to use a variable name other that `fs` (e.g. `fs1`), which is unusual, and can be pretty hard to figure out .

I think the fix here is to have all the "glue code" scoped in it's own anonymous function, so that it doesn't interfere or conflict with user code.

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
Fixes #84 - wraps JS glue code in an anonymous arrow function so that none of the glue code variables leak out and interfere with user code.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Hey @BinaryFissionGames thanks so much for looking into this issue and solving it! I'm not a Javascript expert but what you did makes a lot of sense. And thanks for the great writeup in #84.

I think CI is failing because it uses a GitHub token/secret that apparently can't be made available to builds of forks =/ Seems like there is no straightforward solution yet: https://github.community/t/make-secrets-available-to-builds-of-forks/16166

Will look into whether we can get around this issue, but for now I will run the tests manually and report back!

Alternatively if you think you might be interested in making future contributions we can invite you to the Project Lovelace organization so you can open PRs from regular branches (not forks) until we figure out the CI + secrets issue.

**Comment by ali-ramadhan** [comment]
Tests pass! Although this manual testing approach is not sustainable long term of course.

```
============================================ test session starts =============================================
platform linux -- Python 3.8.5, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /home/alir/bin/miniconda3/bin/python
cachedir: .pytest_cache
rootdir: /home/alir/lovelace-engine, configfile: setup.cfg
collected 6 items                                                                                            

tests/test_javascript_solutions.py::test_javascript_solutions_exist PASSED
tests/test_javascript_solutions.py::test_submit_file[earthquake-epicenters]  2.045713 seconds PASSED
tests/test_javascript_solutions.py::test_submit_file[colorful-resistors]  2.336512 seconds PASSED
tests/test_javascript_solutions.py::test_submit_file[dna-transcription]  2.052197 seconds PASSED
tests/test_javascript_solutions.py::test_submit_file[rocket-science]  2.539968 seconds PASSED
tests/test_javascript_solutions.py::test_submit_file[chaos]  2.626130 seconds PASSED

============================================= 6 passed in 11.63s =============================================
```

I also modified `chaos.js` to add `const fs = require("fs");` at the top of the file and it still passed.

Would probably be useful for debugging to be able to submit a file manually to the engine so I opened an issue to look into this (#87).

**Comment by BinaryFissionGames** [comment]
@ali-ramadhan Yeah, that secrets issue is a real bummer, seems like GHA has no way to have those secrets work in forks. I'd be interested in making more contributions in the future, so if you'd invite me to the org and make it easier on both of us, that'd be great. If you're comfortable with doing that, of course.

Speaking of testing, it'd be nice if there was a way to test as an external contributor locally. Of course, I know you want to keep the problem solutions hidden, but maybe there could be like a subset of problems or maybe even just some dummy test problems/solutions in this repo just for testing locally without any keys. Would probably be less sound than testing with the actual problems/solutions, but more sound than having nothing at all for testing without access to that solution/testing repo.

**Review by ali-ramadhan — APPROVED** [review]

**Comment by ali-ramadhan** [comment]
> I'd be interested in making more contributions in the future, so if you'd invite me to the org and make it easier on both of us, that'd be great. If you're comfortable with doing that, of course.

That would be awesome, thanks so much for your help! Invite sent! You should be able to merge this PR once you accept. (I try not to merge other people's PRs in case they want to make little changes before merging.)

> Speaking of testing, it'd be nice if there was a way to test as an external contributor locally. Of course, I know you want to keep the problem solutions hidden, but maybe there could be like a subset of problems or maybe even just some dummy test problems/solutions in this repo just for testing locally without any keys. Would probably be less sound than testing with the actual problems/solutions, but more sound than having nothing at all for testing without access to that solution/testing repo.

Yeah for sure. With only ~10 problems it makes sense to test all of them but if you have ~100 then a subset could probably test most of the engine's functionality. We mostly just need to test that different return types are treated correctly.

Doing some searching I came across this example repo: https://github.com/imjohnbo/ok-to-test

It seems like they set up a small set of unit tests that run on all PRs, but only run the full tests (with secrets) once a maintainer says it's `ok-to-test`. The CI setup seems more complicated but what do you think of it?

**Comment by BinaryFissionGames** [comment]
I'm not super knowledgeable about GitHub Actions, but that's looks like it solves the problem pretty well, actually. Should add that to the issue list :+1: 

I'll merge this thing now!

**Comment by ali-ramadhan** [comment]
@all-contributors please add @BinaryFissionGames for code and bug

**Pull request merged**
